src/web100srv.c

```diff
@@ -1527,647 +1382,422 @@ void child_process(int parent_pipe, SSL_CTX *ssl_context, int ctlsockfd) {
+    log_println(1, "The child returned! This should never happen.");
+    exit(1);
+  }
 
-      tmp_ptr = mchild;
-
-      // Update queued clients, send message to client when it moves
-      //  up in the queue enough to get closer to running a test.  This
-      //  happens when the client falls into the next lower maxqueue bin
-
-      if (waiting > (2 * max_clients)) {
-        int wait_data = -1;
-        int num_secs = 0;
-        for (i = max_clients; i <= waiting; i++) {
-          if (tmp_ptr == NULL) break;
-          if (i == (2 * max_clients)) {
-            wait_data = 1;
-            num_secs = 45;
-          }
-          if (i == (3 * max_clients)) {
-            wait_data = 2;
-            num_secs = 90;
-          }
-          if (wait_data != -1) {
-            child_signal = wait_data;
-            write(tmp_ptr->pipe, &child_signal, sizeof(int));
-            log_println(6,
-                        "sent %d sec update message to client %d on fd=%d, "
-                        "send_msg() returned %d",
-                        num_secs, tmp_ptr->pid, tmp_ptr->pipe);
-          }
-          tmp_ptr = tmp_ptr->next;
-        }
-      }
-      goto dispatch_client;
-    }
+  // This is the parent process. Handle scheduling and queuing of multiple
+  // incoming clients.
+  log_println(5, "Parent process spawned child = %d", child_pid);
+  log_println(5, "Parent thinks pipe() returned fd0=%d, fd1=%d", child_pipe[0],
+              child_pipe[1]);
+
+  // Close the open resources that should only be used by the child.
+  close(child_pipe[0]);
+  close(ctlsockfd);
+
+  // Initialize the members of new_child
+  new_child = (ndtchild *)calloc(1, sizeof(ndtchild));
+  if (new_child == NULL) {
+    log_println(1, "calloc() failed errno=%d", errno);
+    close(child_pipe[1]);
+    kill(child_pid, SIGKILL);
+    return NULL;
+  }
+  new_child->pid = child_pid;
+  new_child->pipe = child_pipe[1];
+  strlcpy(new_child->addr, rmt_addr, sizeof(new_child->addr));
+  strlcpy(new_child->host, rmt_host, sizeof(new_child->host));
+  new_child->qtime = time(0);
+  new_child->running = 0;
+  new_child->next = NULL;
+
+  return new_child;
+}
 
-    if ((waiting < 0) || (mclients < 0)) {
-      log_println(6,
-                  "Fault: Negative number of clents waiting=%d, mclients=%d"
-                  ", nuke them",
-                  waiting, mclients);
-      while (head_ptr != NULL) {
-        // give message to server waiting in queue
-        child_signal = SRV_QUEUE_SERVER_BUSY;
-        write(head_ptr->pipe, &child_signal, sizeof(int));
-        close(head_ptr->pipe);
-        tpid = head_ptr->pid;
-        child_sig(-1);
-        kill(tpid, SIGTERM);
-        child_sig(tpid);
-      }
-      waiting = 0;
-      mclients = 0;
-      continue;
+/**
+ * Returns true when the current queue size indicates that a new client would
+ * overload the server.
+ */
+int server_is_overloaded(int queue_size) {
```

Comment:
This function name isn't actually accurate because a caller server_is_overloaded will return true if the server is not overloaded. Suggest `server_queue_is_full` or `server_has_queue_capacity`